rfc/20240824-provider-cache-locking.md

```diff
@@ -0,0 +1,68 @@
+# Safe Global Provider Cache Access
+
+Issue: https://github.com/OpenTofu/opentofu/issues/1483
+
+Many CI/CD systems download all providers for every single tofu action run. The global provider cache allows downloaded providers to be saved in between runs and linked into the local provider cache (in .terraform/). Unfortunately, the global provider cache does not have any locking around it and can fail when used in multiple simultaneous runs.
+
+Another common use case is terragrunt. When running `tofu init` on multiple projects simultaneously via the tool, there is a high likely hood of a conflicting write to a provider's directory. This has caused some interesting tooling workarounds on their end, which are not ideal.
+
+Additionally, as we build out true e2e tests for OpenTofu, safe access to the global provider cache can dramatically reduce run times.
+
+## Proposed Solution
+
+A filesystem level lock should be added that is safe for cross-process and in some scenarios cross-machine access. It will be best-effort and should rely on standard locking practices (not home grown) and be transparent to the user.
+
+### User Documentation
+
+By setting `TF_PLUGIN_CACHE_DIR` or `plugin_cache_dir` to a directory, tofu will download all required providers to that directory during init. If providers already exist within that directory, the lock file will be checked and the download may be skipped. Once the global cache is populated, providers will be linked into the local provider cache.  This can save significant time/resources across multiple runs/projects.
+
+Example:
+```
+$ TF_PLUGIN_CACHE_DIR=~/.tofu.d/plugin-cache/ tofu init
+Initializing provider plugins...
+- Finding latest version of hashicorp/local...
+- Installing hashicorp/local v2.5.1...
+
+$ rm .terraform/ -r
+$ TF_PLUGIN_CACHE_DIR=~/.tofu.d/plugin-cache/ tofu init
+Initializing provider plugins...
+- Reusing previous version of hashicorp/local from the dependency lock file
+- Using hashicorp/local v2.5.1 from the shared cache directory
+```
+
+Any number of tofu instances should be able to be run in parallel and still have safe access to the global cache directory.
+
+#### Scenarios which currently cause failures
+
+Multiple init/plans/apply without a primed cache.
+
+One possibility is that each project downloads and overwrites the same provider files.  The global provider cache scan is not a fast process and should be run as little as possible, it is therefore only run at the start of `tofu init`. ProjectA and ProjectB's inits may be run at the same time.  When this occurs, they will both download the full set of providers that they need and clobber each others files.  This is not ideal, but is only a waste of time/resources.
```

Comment:
If we're doing a filesystem level lock, and the files are clobbered. does that mean that we could end up with a clash of locking? I think we should write a bit more information about the impact of this apart from just time/resources and explore possibilities a bit more incase people have these kinds of questions in the future.

Reply:
This is under the section of current failure scenarios, not what we are proposing happen under the new setup